### PR TITLE
Generate JWT token on attest for OptOut and Core

### PR DESCRIPTION
**Pull request by thomasm-ttd**

### CODE REVIEW

File: src/main/java/com/uid2/core/vertx/CoreVerticle.java [hunk 1]
```diff
@@ -220,11 +226,33 @@ private void handleAttestAsync(RoutingContext rc) {
...
                 }
+                responseObj.put("attestation_token", attestationToken);
+                responseObj.put("expiresAt", encryptedAttestationToken.getExpiresAt());
+
+                // TODO: this test should be removed once the KMS work has been done and the KeyId is set.
+                String keyId = ConfigStore.Global.get(Const.Config.AwsKmsJwtSigningKeyIdProp);
+                if (keyId != null && !keyId.isEmpty()) {
+                    try {
+                        String clientKey = "unknown client";
+                        if (rc.request().headers().contains(Const.Http.AppVersionHeader)) {
```
Comment (aulme) [inline src/main/java/com/uid2/core/vertx/CoreVerticle.java:237]: Maybe can extract an `extractClientKey` function from here - the method is getting a bit busy.

File: src/main/java/com/uid2/core/vertx/CoreVerticle.java [hunk 2]
```diff
@@ -335,36 +363,36 @@ JsonObject make(String name, String failReason) {
...
             JsonArray enclaves = (JsonArray) enclavesObj;
-            for (int i=0;i<enclaves.size();i++) {
+            for (int i = 0; i < enclaves.size(); i++) {
                 Result result = new Result();
                 JsonObject item = enclaves.getJsonObject(i);
                 String name = item.getString("name", "__item_" + String.valueOf(i));
                 String proto = item.getString("protocol", null);
                 String identifier = item.getString("identifier", null);
-                if(proto == null) {
+                if (proto == null) {
```
Comment (aulme) [inline src/main/java/com/uid2/core/vertx/CoreVerticle.java:395]: Best rename to protocol, when seeing proto first thought of "prototype" - too much time in JS perhaps.

File: src/main/java/com/uid2/core/vertx/CoreVerticle.java
Comment (aulme) [inline src/main/java/com/uid2/core/vertx/CoreVerticle.java:1]: Feels like there is too much going on in this method/class, perhaps there are some opportunities to logically break it down/extract.

File: src/test/java/com/uid2/core/service/JWTTokenProviderTests.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        contentJson.put("sub", "subject");
+        contentJson.put("iss", "issuer");
+
+        assertJWT(defaultHeaders.encode(), contentJson.encode(), expectedSig, result);
+        assertEquals("1234", this.capturedSignRequest.getValue().keyId());
+        assertEquals(SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256, this.capturedSignRequest.getValue().signingAlgorithm());
+    }
+
+    @Test
+    void getJWT_empty_signature_throws_exception() {
```
Comment (aulme) [inline src/test/java/com/uid2/core/service/JWTTokenProviderTests.java:77]: So far, we're naming tests with camelCase. While I think snake_case would indeed be more readable, this currently would be inconsistent.

File: src/main/java/com/uid2/core/service/JWTTokenProvider.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    public JWTTokenProvider(KmsClientBuilder clientBuilder) {
+        this.kmsClientBuilder = clientBuilder;
+    }
+
+    public String getJWT(Map<String, String> customClaims) throws JwtSigningException {
+        return this.getJWT(null, customClaims);
+    }
+
+    public String getJWT(Map<String, String> headers, Map<String, String> customClaims) throws JwtSigningException {
+        // headers we are going to use are:
```
[the remark this answers is not in the document]
Comment (thomasm-ttd) [reply src/main/java/com/uid2/core/service/JWTTokenProvider.java:43]: removed

File: src/test/java/com/uid2/core/service/JWTTokenProviderTests.java (hunk 3 above)
Comment (thomasm-ttd) [reply src/test/java/com/uid2/core/service/JWTTokenProviderTests.java:85]: Updated

File: src/main/java/com/uid2/core/vertx/CoreVerticle.java
Comment (thomasm-ttd) [reply src/main/java/com/uid2/core/vertx/CoreVerticle.java:1]: refactored

File: src/main/java/com/uid2/core/service/OperatorJWTTokenProvider.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        claims.put("aud", audience);
+        claims.put("roles", roleString);
+        claims.put("siteId", siteId.toString());
+        claims.put("enclaveId", enclaveId);
+        claims.put("enclaveType", enclaveType);
+        claims.put("operatorVersion", operatorVersion);
+
+        LOGGER.debug(String.format("Creating token with: Issuer: %s, Audience: %s, Roles: %s, SiteId: %s, EnclaveId: %s, EnclaveType: %s, OperatorVersion: %s", audience, issuer, roleString, siteId, enclaveId, enclaveType, operatorVersion));
+        String token = this.jwtTokenProvider.getJWT(expiresAt, this.clock.instant(), claims);
+        return token;
```
Comment (aulme) [inline src/main/java/com/uid2/core/service/OperatorJWTTokenProvider.java:77]: inline?

File: src/main/java/com/uid2/core/service/JWTTokenProvider.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            SignRequest request = SignRequest.builder()
+                    .keyId(keyId)
+                    .signingAlgorithm(SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256)
+                    .message(dataToSign)
+                    .build();
+
+            SignResponse response = kmsClient.sign(request);
+            if (response.sdkHttpResponse().isSuccessful()) {
+                String signature = encoder.encodeToString(response.signature().asByteArray());
+                return signature;
```
Comment (aulme) [inline src/main/java/com/uid2/core/service/JWTTokenProvider.java:99]: inline?

**Review by aulme — APPROVED** [review]

**Pull request merged**